x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -727,16 +814,313 @@ public void runAsDenied(String requestId, Authentication authentication, RestReq
+                builder.field("query", indicesPrivileges.getQuery().utf8ToString());
+            }
+            // default for "allow_restricted_indices" is false, and it's very common to stay that way, so don't show it unless true
+            if (indicesPrivileges.allowRestrictedIndices()) {
+                builder.field("allow_restricted_indices", indicesPrivileges.allowRestrictedIndices());
+            }
+            builder.endObject();
+        }
+
+        LogEntryBuilder withRequestBody(DeleteUserRequest deleteUserRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "delete_user");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("user")
+                        .field("name", deleteUserRequest.username())
+                    .endObject() // user
+                    .endObject();
+            logEntry.with(DELETE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(DeleteRoleRequest deleteRoleRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "delete_role");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("role")
+                    .field("name", deleteRoleRequest.name())
+                    .endObject() // role
+                    .endObject();
+            logEntry.with(DELETE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(DeleteRoleMappingRequest deleteRoleMappingRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "delete_role_mapping");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("role_mapping")
+                    .field("name", deleteRoleMappingRequest.getName())
+                    .endObject() // role_mapping
+                    .endObject();
+            logEntry.with(DELETE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(InvalidateApiKeyRequest invalidateApiKeyRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "invalidate_apikeys");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("apikeys");
+                if (invalidateApiKeyRequest.getIds() != null && invalidateApiKeyRequest.getIds().length > 0) {
+                    builder.array("ids", invalidateApiKeyRequest.getIds());
+                }
+                if (Strings.hasLength(invalidateApiKeyRequest.getName())) {
```

Comment:
Similar situation here: I think the issue is we are using `Strings#hashLength` here. This means a whitespace string, e.g. `"   
 "` will be output literally here, while it is actually treated as not-specified in the code (tested with `Strings#hasText`). I'd prefer we log consistently for empty (equivalent to not-specified) values.

Reply:
A moot point by now, but the common case is for the field to be `null` in which case we don't print it.
But, if it is the blank string, it was set explicitly by the user, which might be exploiting some bug.
In principle, moving the logic from the transport handler to the audit formatter is brittle.